File: C++/sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {

        if (head == nullptr) {
            return nullptr; // empty
        }
        
        if (head->next == nullptr) {
            return head; // 1 element, already sorted
        }
        
        // split in half
        
        ListNode* slow = head;
        ListNode* fast = head->next;
        
        while (fast) {
            if (fast->next) {
                fast = fast->next->next;
            } else {
                break;
            }
            slow = slow->next;
        }
        
        ListNode* a = head;
        ListNode* b = slow->next;
        
        slow->next = nullptr;
        
        // sort halves
        
        a = sortList(a);
        b = sortList(b);
        
        // merge
        
        ListNode* ans = merge(a, b);

        return ans;
    }
    
    ListNode* merge(ListNode* a, ListNode* b) {
        if (a && b) {
            if (a->val < b->val) {
                a->next = merge(a->next, b);
                return a;
            } else {
                b->next = merge(a, b->next);
                return b;
            }
        } else if (a) {
            return a;
        }
        return b;
    }
};

File: Python3/sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        values = defaultdict(lambda: 0)
        node = head
        while node:
            values[node.val] += 1
            node = node.next
        result = None
        for k, v in sorted(values.items(), reverse=True):
            for _ in range(v):
                result = ListNode(k, result)
        return result

